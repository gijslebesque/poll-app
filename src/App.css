.App {
	/* max-width: 100vw; */
	overflow: hidden;
}

h1 {
	font-size: 40px;
}
header {
	text-align: center;
}

aside {
	padding: 3rem 1.5rem;
}
nav {
	padding: 1.5rem 1.5rem;
}

.icon.is-small.is-left {
	top: 50%;
	transform: translateY(-50%);
}

.flex-container {
	display: flex;
	justify-content: space-between;
}

.flex-container input {
	width: 80%;
}
#notification-root {
	position: fixed;
	top: 10px;
	right: 10px;
	overflow: hidden;
	z-index: 9;
}
.notification {
	position: fixed;
	top: 20px;
	right: 20px;
	max-width: 250px;
	transition: 1s all;
	z-index: 9999;
}
.is-primary {
	margin: 20px 20px 20px 0px;
}

.navbar-menu.is-active {
	position: absolute;
	left: 20px;
	right: 20px;
	z-index: 9;
}

.modal-content {
	background: white;
	padding: 40px;
	border-radius: 10px;
}

input[type="radio"] {
	margin-right: 5px;
}
.chart-container {
	text-align: center;
	width: 320px;
	height: 250px;
	transform: translateX(-15%);
}

@media (min-width: 1025px) {
	.chart-container {
		width: 500px;
		height: 350px;
	}
}
