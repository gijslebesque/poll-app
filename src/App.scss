.App {
	overflow: hidden;
}

h1 {
	font-size: 40px;
}

header {
	text-align: center;
}

aside {
	padding: 3rem 1.5rem;
}

nav {
	padding: 1.5rem 1.5rem;
}

.icon.is-small.is-left {
	top: 50%;
	transform: translateY(-50%);
}

.flex-container {
	display: flex;
	justify-content: space-between;
	input {
		width: 80%;
	}
}

.notification {
	position: fixed;
	top: 20px;
	right: 20px;
	max-width: 250px;
	transition: 1s all;
	z-index: 9999;
}

.is-primary {
	margin: 20px 20px 20px 0px;
}

.navbar-menu.is-active {
	position: absolute;
	left: 20px;
	right: 20px;
	z-index: 9;
}

.modal-content {
	background: white;
	padding: 40px;
	border-radius: 10px;
}

input[type="radio"] {
	margin-right: 5px;
}

.chart-container {
	text-align: center;
	width: 320px;
	height: 250px;
	transform: translateX(-15%);
}

@media (min-width: 1025px) {
	.chart-container {
		width: 500px;
		height: 350px;
		transform: translateX(0%);
	}
}
