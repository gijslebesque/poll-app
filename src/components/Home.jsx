import React from "react";
import Section from "./Section";

const Home = () => {
	return (
		<Section title="Welcome">
			<p>
				Ullamco et fugiat tempor esse ex. Exercitation laboris ex dolore dolor
				cupidatat labore occaecat consectetur fugiat. Occaecat occaecat occaecat
				tempor aliquip velit magna Lorem officia labore. Velit consectetur minim
				qui amet irure reprehenderit quis elit anim. Lorem magna nulla fugiat
				consectetur deserunt incididunt est officia occaecat nulla adipisicing
				deserunt ullamco ut. Mollit minim officia non sint aliqua aliquip
				voluptate. Culpa excepteur dolore commodo anim magna pariatur sint
				officia exercitation est consectetur aliqua excepteur.
			</p>
		</Section>
	);
};

export default Home;
