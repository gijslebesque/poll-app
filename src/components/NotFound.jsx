import React from "react";
import Section from "./Section";
import history from "../history";
const NotFound = () => {
	return (
		<Section title="404 Not found" subtitle="Sorry there's nothing here">
			<button className="button is-primary" onClick={() => history.push("/")}>
				Back home
			</button>
		</Section>
	);
};

export default NotFound;
