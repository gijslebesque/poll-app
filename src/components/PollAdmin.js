import React, { Component } from "react";
import PropTypes from "prop-types";
import Form from "./Form.jsx";
import Section from "./Section";
import Notification from "./Notification";

class PollAdmin extends Component {
	constructor(props) {
		super(props);
		this.state = {
			edit: false,
			style: {
				transform: "translate(100%)",
				transition: "all 1s ease"
			},
			poll: {
				question: "",
				answers: [
					{
						name: 0,
						value: ""
					},
					{
						name: 0,
						value: ""
					}
				]
			},
			notAllowedNotification: false
		};
	}

	isEdit = () => {
		const id = this.props.match.params.id;
		if (id) {
			const poll = this.props.getPoll(id);
			if (
				poll[0] &&
				(poll[0].owner === this.props.userName || poll[0].owner === "Anonymous")
			)
				this.setState({ poll: poll[0], edit: true });
		}
	};

	componentDidMount() {
		setTimeout(this.mountStyle, 10);
		this.isEdit();
	}

	componentDidUpdate(prevProps) {
		if (prevProps.userName !== this.props.userName) {
			this.isEdit();
		}
	}

	clickNextHandler = e => {
		this.unMountStyle();
		setTimeout(() => {
			this.props.clickNextHandler(e);
		}, 200);
	};

	mountStyle = () => {
		this.setState({
			style: {
				transform: "translate(0%)",
				transition: "all 0.5s ease"
			}
		});
	};

	unMountStyle = () => {
		// css for mount animation
		this.setState({
			style: {
				transform: "translate(-100%)",
				opacity: 1,
				transition: "all 0.5s ease"
			}
		});
	};

	handleQuestion = e => {
		const poll = { ...this.state.poll };
		poll.question = e.target.value;
		this.setState({ poll });
	};

	addQuestion = () => {
		if (this.state.poll.answers.length >= 10) return;
		const newInput = {
			name: this.state.poll.answers.length,
			value: ""
		};
		this.setState(prevState => ({
			poll: {
				...prevState.poll,
				answers: prevState.poll.answers.concat([newInput])
			}
		}));
	};

	changeQuestion = (e, i) => {
		e.preventDefault();
		const answers = [...this.state.poll.answers];
		answers[i].value = e.target.value;

		this.setState(prevState => ({
			poll: {
				...prevState.poll,
				answers: answers
			}
		}));
	};

	deleteInput = i => {
		const answers = [...this.state.poll.answers];
		answers.splice(i, 1);
		this.setState(prevState => ({ poll: { ...prevState.poll, answers } }));
	};

	showNotAllowed = () => {
		if (this.state.notAllowedNotification) return;
		this.setState({ notAllowedNotification: true });
		setTimeout(() => this.setState({ notAllowedNotification: false }), 3000);
	};

	addPoll = () => {
		const answers = this.state.poll.answers;

		for (let i = answers.length - 1; i > 0; i--) {
			if (!answers[i].value) {
				this.showNotAllowed();
				return;
			}
		}

		if (this.state.poll.answers.length < 2 || !this.state.poll.question) {
			this.showNotAllowed();
		} else {
			const id = this.props.match.params.id;
			this.props.addPoll(this.state.poll, id);
		}
	};
	render() {
		const copy = this.state.edit ? "Edit" : "Add";

		const disabled =
			this.state.poll.answers.length >= 10 ? { disabled: "disabled" } : "";
		return (
			<>
				<Section
					style={this.state.style}
					title={`${copy} your poll`}
					subtitle="Enter a question and add up to 10 possible answers."
				>
					<h3>Question</h3>
					<Form>
						<input
							tabIndex="1"
							className="input"
							type="text"
							placeholder="What do you want to poll?"
							name="question"
							value={this.state.poll.question}
							onChange={e => this.handleQuestion(e)}
						/>
					</Form>
					<h3>Answers</h3>
					{this.state.poll.answers.map((input, i) => (
						<Form key={i}>
							<div className="flex-container" key={i}>
								<input
									tabIndex={i + 2}
									className="input"
									type="text"
									placeholder="Provide an answer"
									name={input.name}
									value={input.value}
									onChange={e => {
										this.changeQuestion(e, i);
									}}
								/>
								<button
									className="button is-link"
									onClick={() => this.deleteInput(i)}
								>
									Delete
								</button>
							</div>
						</Form>
					))}

					<button
						className="button is-outlined is-primary"
						onClick={this.addQuestion}
						{...disabled}
					>
						Add an answer
					</button>

					<button className="button is-primary" onClick={this.addPoll}>
						{`${copy} your poll`}
					</button>

					{this.state.edit && (
						<button
							className="button is-primary"
							onClick={() => this.props.deletePoll(this.props.match.params.id)}
						>
							Delete this poll
						</button>
					)}

					<p>{this.state.poll.answers.length}/10 possible answers.</p>
					{!this.props.userName && (
						<Notification
							type="is-warning"
							text="You are currently not logged in. You can still create a poll, but it can be edited by anyone. Please login to make it personal."
						/>
					)}

					{this.state.notAllowedNotification && (
						<Notification
							type="is-danger"
							text="You need to add at least two answers. And a question."
						/>
					)}
				</Section>
			</>
		);
	}
}

PollAdmin.propTypes = {
	getPoll: PropTypes.func.isRequired,
	addPoll: PropTypes.func.isRequired,
	deletePoll: PropTypes.func.isRequired,
	userName: PropTypes.string.isRequired
};

export default PollAdmin;
