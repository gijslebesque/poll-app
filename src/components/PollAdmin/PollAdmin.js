import React, { Component } from "react";
import PropTypes from "prop-types";
import Form from "../Form.jsx";
import Section from "../Section";
import Notification from "../Notification";

class PollAdmin extends Component {
	constructor(props) {
		super(props);
		this.state = {
			style: {
				transform: "translate(-100%)",
				transition: "all 1s ease"
			},
			poll: {
				question: "",
				answers: [
					{
						name: 0,
						value: ""
					},
					{
						name: 0,
						value: ""
					}
				]
			}
		};
	}

	componentDidMount() {
		setTimeout(this.mountStyle, 10);
	}

	clickNextHandler = e => {
		this.unMountStyle();
		setTimeout(() => {
			this.props.clickNextHandler(e);
		}, 200);
	};

	mountStyle = () => {
		this.setState({
			style: {
				transform: "translate(0%)",
				transition: "all 0.5s ease"
			}
		});
	};

	unMountStyle = () => {
		// css for mount animation
		this.setState({
			style: {
				transform: "translate(-100%)",
				opacity: 1,
				transition: "all 0.5s ease"
			}
		});
	};

	handleQuestion = e => {
		console.log(e.target.value);
		debugger;
		const poll = { ...this.state.poll };
		poll.question = e.target.value;
		this.setState({ poll });
		// this.setState(prevState => ({
		// 	poll: { ...prevState.poll, question: e.target.value }
		// }));
	};

	addQuestion = e => {
		if (this.state.poll.answers.length >= 10) return false;
		e.preventDefault();
		var newInput = {
			name: this.state.poll.answers.length,
			value: ""
		};
		this.setState(prevState => ({
			poll: {
				...prevState.poll,
				answers: prevState.poll.answers.concat([newInput])
			}
		}));
	};

	changeQuestion = (e, i) => {
		e.preventDefault();
		const answers = [...this.state.poll.answers];
		answers[i].value = e.target.value;

		this.setState(prevState => ({
			poll: {
				...prevState.poll,
				answers: answers
			}
		}));
	};

	deleteInput = (e, i) => {
		e.preventDefault();
		if (this.state.answers.length <= 2) return false;
		const answers = [...this.state.answers];
		answers.splice(i, 1);
		this.setState({ answers });
	};

	render() {
		return (
			<Section
				style={this.state.style}
				title="Create your poll"
				subtitle="Fill in a question and add up to 10 possible answers to it."
			>
				<Form>
					<input
						className="input"
						type="text"
						placeholder="What do you want to poll?"
						name="question"
						onChange={e => this.handleQuestion(e)}
					/>
				</Form>

				{this.state.poll.answers.map((input, i) => (
					<Form>
						<div className="flex-container" key={i}>
							<input
								className="input"
								type="text"
								placeholder="Provide an answer"
								name={input.name}
								value={input.value}
								onChange={e => {
									this.changeQuestion(e, i);
								}}
							/>
							<button
								className="button is-link"
								onClick={e => this.deleteInput(e, i)}
							>
								Delete
							</button>
						</div>
					</Form>
				))}

				<button className="button" onClick={e => this.addQuestion(e)}>
					Add an answer
				</button>
				<button
					className="button"
					onClick={e => this.props.addPoll(this.state.poll)}
				>
					Add your poll
				</button>
				<Notification type="is-warning" />
			</Section>
		);
	}
}

PollAdmin.propTypes = {
	clickNextHandler: PropTypes.func.isRequired,
	handleQuestion: PropTypes.func.isRequired,
	answers: PropTypes.array.isRequired,
	changeQuestion: PropTypes.func.isRequired,
	deleteInput: PropTypes.func.isRequired,
	addQuestion: PropTypes.func.isRequired
};

export default PollAdmin;
