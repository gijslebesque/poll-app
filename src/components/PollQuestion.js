import React, { Component } from "react";
import PollAnswers from "./PollChart";
import PropTypes from "prop-types";
import Section from "./Section";
import Form from "./Form.jsx";

class PollQuestions extends Component {
	constructor(props) {
		super(props);
		this.state = {
			checked: "",
			showEditModal: false,
			userAnswers: [],
			style: {
				transform: "translate(100%)"
			},
			poll: {
				question: "",
				answers: []
			}
		};
	}
	componentDidMount() {
		// call the animation
		this.showPoll();
		setTimeout(this.mountStyle, 10);
	}

	componentDidUpdate(prevProps) {
		const {
			match: {
				params: { id }
			}
		} = this.props;

		if (prevProps.match.params.id !== id) {
			this.showPoll();
		}
	}

	mountStyle = () => {
		// css for mount animation
		this.setState({
			style: {
				transform: "translate(0%)",
				transition: "all 0.5s ease"
			}
		});
	};
	unMountStyle = () => {
		// css for unmount animation
		this.setState({
			style: {
				transform: "translate(100%)",
				transition: "all 0.5s ease"
			}
		});
	};

	vote = () => {
		const updatedUserAnswers = this.state.userAnswers.map((answer, i) => {
			if (i === this.state.checked) {
				answer.vote++;
			}
			return answer;
		});
		this.setState({ userAnswers: updatedUserAnswers, checked: "" });
	};
	reset = () => {
		this.showPoll();
	};
	edit = () => {
		this.props.history.push({
			pathname: `/create-poll/${this.props.match.params.id}`,
			poll: this.state.poll
		});
	};
	showPoll = () => {
		const id = this.props.match.params.id;
		const polls = this.props.polls;

		const poll = polls.filter((poll, i) => {
			return i === parseInt(id);
		});

		if (!poll[0]) {
			this.props.history.push("/");
			return;
		}
		const userAnswers = poll[0].answers.map((answer, i) => {
			return {
				name: `Answer ${i + 1}`,
				vote: 0
			};
		});

		this.setState({ poll: poll[0], userAnswers });
	};
	handleRadioBtn = e => {
		const indexBtn = parseInt(e.target.getAttribute("data-index"));
		this.setState({ checked: indexBtn });
	};

	render() {
		const { poll } = this.state;
		return (
			<div style={this.state.style}>
				<Section title={poll.question} subtitle={`asked by ${poll.owner}`}>
					<div className="columns">
						<div className="column">
							<Form name="Answers">
								{poll.answers.map((answer, i) => {
									return (
										<div className="control" key={i}>
											<label className="radio">
												<input
													type="radio"
													name="answer"
													checked={this.state.checked === i}
													data-index={i}
													value={answer.value}
													onChange={e => this.handleRadioBtn(e)}
												/>
												{answer.value}
											</label>
										</div>
									);
								})}
								<button className="button is-primary" onClick={this.vote}>
									Vote!
								</button>
								<button className="button is-primary" onClick={this.reset}>
									Reset
								</button>
								{(this.state.poll.owner === this.props.userName ||
									this.state.poll.owner === "Anonymous") && (
									<button className="button is-primary" onClick={this.edit}>
										edit
									</button>
								)}
							</Form>
						</div>
						<div className="column">
							<PollAnswers userAnswers={this.state.userAnswers} />
						</div>
					</div>
				</Section>
			</div>
		);
	}
}

PollQuestions.propTypes = {
	polls: PropTypes.array.isRequired
};

PollQuestions.defaultProps = {
	polls: []
};

export default PollQuestions;
