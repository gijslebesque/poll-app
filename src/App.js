import React, { Component } from "react";
import "bulma/css/bulma.css";
import NavBar from "./components/NavBar";
import "./App.css";
import { Route, Switch } from "react-router-dom";
import PollAdmin from "./components/PollAdmin";
import PollQuestion from "./components/PollQuestion";
import Modal from "./components/Modal";
import history from "./history";
import Menu from "./components/Menu";
import Home from "./components/Home";
import NotFound from "./components/NotFound";
import ErrorBoundry from "./components/ErrorBoundry";

class App extends Component {
	constructor() {
		super();
		this.state = {
			userName: "",
			showModal: false,
			polls: [
				{
					owner: "Gijs",
					question: "Hello there how can I help you?",
					answers: [
						{
							name: 0,
							value:
								"One Sint magna officia commodo sint quis cupidatat veniam dolore cillum labore reprehenderit."
						},
						{
							name: 0,
							value: "Two"
						},
						{
							name: 0,
							value: "Three"
						}
					]
				}
			]
		};
	}

	getPoll = id => {
		const stateCopy = [...this.state.polls];
		return stateCopy.splice(id, 1);
	};
	setUserName = (field, name) => {
		this.setState({ [field]: name, showModal: false });
	};

	logout = () => {
		this.setState({ userName: "" });
		history.push("/");
	};
	toggleModal = () => {
		this.setState({ showModal: !this.state.showModal });
	};

	addPoll = (poll, id) => {
		const polls = [...this.state.polls];
		poll.owner = this.state.userName ? this.state.userName : "Anonymous";
		if (id) {
			polls[id] = poll;
			history.push(`/poll/${id}`);
		} else {
			polls.push(poll);
			history.push(`/poll/${polls.length - 1}`);
		}
		this.setState({ polls });
	};

	deletePoll = index => {
		const polls = [...this.state.polls];
		polls.splice(parseInt(index), 1);
		this.setState({ polls });
		history.push("/");
	};

	render() {
		return (
			<div className="App">
				<NavBar
					userName={this.state.userName}
					toggleModal={this.toggleModal}
					logout={this.logout}
				/>
				<>
					<header>
						<h1> {`Sir ${this.state.userName} Poll-a-lot`} </h1>
						<h2>Poll some</h2>
					</header>
					<div className="columns">
						<div className="column">
							<Menu polls={this.state.polls} />
						</div>
						<div className="column is-four-fifths">
							<Switch>
								<Route exact path="/" component={Home} />
								<Route
									path="/create-poll/:id?"
									render={props => (
										<ErrorBoundry>
											<PollAdmin
												{...props}
												getPoll={this.getPoll}
												userName={this.state.userName}
												addPoll={this.addPoll}
												deletePoll={this.deletePoll}
											/>
										</ErrorBoundry>
									)}
								/>
								<Route
									exact
									path="/poll/:id"
									render={props => (
										<PollQuestion
											{...props}
											userName={this.state.userName}
											polls={this.state.polls}
										/>
									)}
								/>
								<Route path="/*" component={NotFound} />
							</Switch>
							{this.state.showModal && (
								<Modal
									setUserName={this.setUserName}
									toggleModal={this.toggleModal}
								/>
							)}
						</div>
					</div>
				</>
			</div>
		);
	}
}

export default App;
