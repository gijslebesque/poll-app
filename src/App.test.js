import React from "react";
import App from "./App";
import { Router, MemoryRouter } from "react-router-dom";
import { shallow, mount, render } from "enzyme";
import Home from "./components/Home";
import NotFound from "./components/NotFound";
import PollAdmin from "./components/PollAdmin";
import PollQuestion from "./components/PollQuestion";

import history from "./history";

describe("Router", () => {
	it("Router should render correctly", () => {
		const component = render(<Router history={history} />);
		expect(component).toMatchSnapshot();
	});

	it("App should render correctly", () => {
		const component = shallow(<App />);
		expect(component).toMatchSnapshot();
	});

	it("Home should render correctly", () => {
		const component = mount(
			<MemoryRouter initialEntries={["/"]}>
				<App />
			</MemoryRouter>
		);

		expect(component.find(Home)).toHaveLength(1);
		expect(component.find(PollAdmin)).toHaveLength(0);
		expect(component.find(PollQuestion)).toHaveLength(0);
		expect(component.find(NotFound)).toHaveLength(0);
	});

	it("PollQuestion should redirect to home correctly", () => {
		const component = mount(
			<MemoryRouter initialEntries={["/poll/0"]}>
				<App />
			</MemoryRouter>
		);
		expect(component.find(Home)).toHaveLength(1);
		expect(component.find(PollAdmin)).toHaveLength(0);
		expect(component.find(PollQuestion)).toHaveLength(0);
		expect(component.find(NotFound)).toHaveLength(0);
	});

	it("NotFound should render correctly", () => {
		const component = mount(
			<MemoryRouter initialEntries={["/random"]}>
				<App />
			</MemoryRouter>
		);

		expect(component.find(Home)).toHaveLength(0);
		expect(component.find(PollAdmin)).toHaveLength(0);
		expect(component.find(PollQuestion)).toHaveLength(0);
		expect(component.find(NotFound)).toHaveLength(1);
	});
});
