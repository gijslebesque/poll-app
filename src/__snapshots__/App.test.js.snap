// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router App should render correctly 1`] = `ShallowWrapper {}`;

exports[`Router Router should render correctly 1`] = `
initialize {
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
